File: pycore/pyfoundations/service_contract.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

SERVICE_CONTRACT_PATH = Path(__file__).resolve().parents[2] / "config" / "service_contract.json"
SERVICE_CONTRACT: dict[str, Any] = json.loads(SERVICE_CONTRACT_PATH.read_text(encoding="utf-8"))


def value(contract_path: str) -> Any:
    current: Any = SERVICE_CONTRACT

    for segment in contract_path.split("."):
        if not isinstance(current, dict) or segment not in current:
            raise KeyError(f"Unknown service contract value: {contract_path}")
        current = current[segment]

    return current


def host(name: str) -> str:
    resolved = value(f"hosts.{name}")
    if not isinstance(resolved, str) or not resolved:
        raise ValueError(f"Invalid service contract host: {name}")
    return resolved


def port(name: str) -> int:
    resolved = value(f"ports.{name}")
    if not isinstance(resolved, int) or resolved < 1:
        raise ValueError(f"Invalid service contract port: {name}")
    return resolved


def root_domain(index: int = 0) -> str:
    domains = value("access.root_domains")
    if not isinstance(domains, list) or index >= len(domains) or not isinstance(domains[index], str):
        raise ValueError(f"Invalid service contract root domain index: {index}")
    return domains[index]


def service_domain(name: str, replacements: dict[str, str] | None = None, root_domain_index: int = 0) -> str:
    labels = value(f"access.service_domains.{name}")
    resolved_replacements = replacements or {}
    default_region = value("access.default_api_region_prefix")
    resolved_labels: list[str] = []

    if not isinstance(labels, list) or not labels:
        raise ValueError(f"Invalid service contract domain: {name}")

    for label in labels:
        if not isinstance(label, str) or not label:
            raise ValueError(f"Invalid service contract domain label: {name}")
        if label.startswith("{") and label.endswith("}"):
            replacement_key = label[1:-1]
            resolved_labels.append(resolved_replacements.get(replacement_key, default_region))
        else:
            resolved_labels.append(label)

    return ".".join([*resolved_labels, root_domain(root_domain_index)])


def build_url(protocol: str, hostname: str, port_number: int | None = None, path: str = "") -> str:
    port_part = f":{port_number}" if port_number else ""
    path_part = f"/{path}" if path and not path.startswith("/") else path
    return f"{protocol}://{hostname}{port_part}{path_part}"

File: config/service_contract.js
const fs = require('fs');
const path = require('path');

const SERVICE_CONTRACT_PATH = path.join(__dirname, 'service_contract.json');
const SERVICE_CONTRACT = JSON.parse(fs.readFileSync(SERVICE_CONTRACT_PATH, 'utf8'));

function value(contractPath) {
  const segments = contractPath.split('.');
  let current = SERVICE_CONTRACT;

  for (const segment of segments) {
    current = current && Object.prototype.hasOwnProperty.call(current, segment)
      ? current[segment]
      : undefined;
  }

  if (current === undefined || current === null) {
    throw new Error(`Unknown service contract value: ${contractPath}`);
  }

  return current;
}

function host(name) {
  return String(value(`hosts.${name}`));
}

function port(name) {
  const resolvedPort = value(`ports.${name}`);

  if (!Number.isInteger(resolvedPort) || resolvedPort < 1) {
    throw new Error(`Invalid service contract port: ${name}`);
  }

  return resolvedPort;
}

function rootDomain(index = 0) {
  const domains = value('access.root_domains');
  const resolvedDomain = Array.isArray(domains) ? domains[index] : undefined;

  if (typeof resolvedDomain !== 'string' || resolvedDomain.length === 0) {
    throw new Error(`Invalid service contract root domain index: ${index}`);
  }

  return resolvedDomain;
}

function serviceDomain(name, replacements = {}, rootDomainIndex = 0) {
  const labels = value(`access.service_domains.${name}`);
  const resolvedLabels = Array.isArray(labels)
    ? labels.map((label) => String(label).replace(/^\{(.+)\}$/, (_, key) => replacements[key] || value(`access.default_api_region_prefix`)))
    : [];

  if (resolvedLabels.length === 0 || resolvedLabels.some((label) => label.length === 0)) {
    throw new Error(`Invalid service contract domain: ${name}`);
  }

  return [...resolvedLabels, rootDomain(rootDomainIndex)].join('.');
}

function url(protocol, hostname, portNumber, pathname = '') {
  const portPart = portNumber ? `:${portNumber}` : '';
  const pathPart = pathname && !pathname.startsWith('/') ? `/${pathname}` : pathname;

  return `${protocol}://${hostname}${portPart}${pathPart}`;
}

module.exports = {
  document: SERVICE_CONTRACT,
  host,
  port,
  rootDomain,
  serviceDomain,
  url,
  value,
};
